apps/prairielearn/src/ee/lib/validateHTML.ts

```diff
@@ -0,0 +1,616 @@
+import * as parse5 from 'parse5';
+
+const HTML_GLOBAL_TAGS = [
+  'accesskey',
+  'class',
+  'contenteditable',
+  'data-*',
+  'dir',
+  'draggable',
+  'enterkeyhint',
+  'hidden',
+  'id',
+  'inert',
+  'inputmode',
+  'lang',
+  'popover',
+  'spellcheck',
+  'style',
+  'tabindex',
+  'title',
+  'translate',
+];
+
+/**
+ * Checks that the required property is an int (or property to be inserted) or adds an error to the provided list.
+ * @param tag The name of the tag being checked.
+ * @param key The property name.
+ * @param val The property value.
+ * @param errors The list of errors to add to.
+ */
+function assertInt(tag: string, key: string, val: string, errors: string[]) {
+  if (!(/^\d+$/.test(val) || /^\{\{.*\}\}$/.test(val))) {
+    errors.push(
+      `${tag}: value for attribute ${key} must be an integer, but value provided is "${val}"`,
+    );
+  }
+}
+
+/**
+ * Checks that the required property is an float (or property to be inserted) or adds an error to the provided list.
+ * @param tag The name of the tag being checked.
+ * @param key The property name.
+ * @param val The property value.
+ * @param errors The list of errors to add to.
+ */
+function assertFloat(tag: string, key: string, val: string, errors: string[]) {
+  if (!(/^(\d+)\.?(\d*)(e-\d+)?$/.test(val) || /^\{\{.*\}\}$/.test(val))) {
+    errors.push(
+      `${tag}: value for attribute ${key} must be an floating-point number, but value provided is "${val}"`,
+    );
+  }
+}
+
+/**
+ * Checks that the required property is in a list of possibilities (or property to be inserted) or adds an error to the provided list.
+ * @param tag The name of the tag being checked.
+ * @param key The property name.
+ * @param val The property value.
+ * @param choices The list of potential choices for the property.
+ * @param errors The list of errors to add to.
+ */
+function assertInChoices(
+  tag: string,
+  key: string,
+  val: string,
+  choices: string[],
+  errors: string[],
+) {
+  if (!(choices.includes(val) || /^\{\{.*\}\}$/.test(val))) {
+    errors.push(
+      `${tag}: value for attribute ${key} must be in ${choices}, but value provided is ${val}`,
+    );
+  }
+}
+
+/**
+ * Checks that the required property is a boolean (or property to be inserted) or adds an error to the provided list.
+ * @param tag The name of the tag being checked.
+ * @param key The property name.
+ * @param val The property value.
+ * @param choices The list of potential choices for the property.
+ * @param errors The list of errors to add to.
+ */
+function assertBool(tag: string, key: string, val: string, errors: string[]) {
+  assertInChoices(tag, key, val, ['true', 'false'], errors);
```

Comment:
Note: this list is incomplete, but I'm not sure if it matters for the sake of the LLM. I think I'd be happy to enforce that it uses the "normal" boolean values.

https://github.com/PrairieLearn/PrairieLearn/blob/ab0010dc0a461a47f78f1e33412e8a90964a4b1e/apps/prairielearn/python/prairielearn.py#L608-L621

Reply:
I wanted to write an optimistic validator, so changed in 0e87e7e4333c2d4d8dfbc2116aad0c0acd81707b.